### DISCUSSION TITLE
Asynchronous Errors in Optimization with Windows Dakota:

### DISCUSSION THREAD

**Discussion by mlaudu** [snl-dakota/dakota#136] [Q&A]
I tried to run an optimization on a Windows machine with Dakota 6.19 installed. The computer is a desktop with 16 cores. If I don't invoke
'asynchronous' in my interface, the optimization runs serially but it converges after a couple of hours. I want to be able to run a bit faster; but I allow 'asynchronous' with no evaluation_concurrency or with evaluation_concurrency = 16, the program aborts and eports an error as shown below:

-----------------------
C:\Users\XXXX\AppData\Local\Temp\dakota_work_c96414fe.14>echo off
Initiating evaluation 15
start "SystemInterface-Evaluation" /B simulator_script.bat params.in results.out

C:\Users\XXXX\AppData\Local\Temp\dakota_work_d3335350.15>echo off
Initiating evaluation 16
start "SystemInterface-Evaluation" /B simulator_script.bat params.in results.out

C:\Users\XXXX\AppData\Local\Temp\dakota_work_6390b8aa.16>echo off

C:\Users\XXXX\AppData\Local\Temp\dakota_work_f0b03b2d.1>Second pass: testing for completions (16 running)

Error: could not remove path "C:\Users\XXXX\AppData\Local\Temp\dakota_work_f0b03b2d.1";
boost::filesystem::remove: The process cannot access the file because it is being used by another process: "C:\Users\XXXX\AppData\Local\Temp\dakota_work_f0b03b2d.1"
-----------------------

I attach here copies of my Dakota input file (dakota.txt) and simulator script (simulator_script.bat.txt). I'll appreciate any advise on how to fix this problem.

[dakota.txt](https://github.com/user-attachments/files/16910443/dakota.txt)
[simulator_script.txt](https://github.com/user-attachments/files/16910444/simulator_script.txt)

**Comment by briadam** [comment]
The system interface detects evaluation completion by the presence of the `results.out` file. It's possible you're getting a race condition where the Dakota collects the `results.out`, but some processes started by `simulator_script.bat` haven't yet terminated.

One potential workaround (simpler that deep debugging) might be to specify `directory_save` such that Dakota won't try to remove the run directories as they complete. Then you could remove them later when the Dakota study is finished. I realize this might not be practical, but might help validate this theory and get you going with a faster study...

**Reply by mlaudu** [reply]
Hi Brian,
Thank you very much for that advise. Yes, you are correct. if I specify the directory_save option, the problem solves; but as you said, I had to deal with quite an inordinate number of folders that were created and had to be deleted later. I appreciate your insight. I'll have to bear with that for now. Cheers.

**Reply by mlaudu** [reply]
My apologies. Maybe I talked too fast because I tested it with 100 function evaluations. When I increased the number to 500, I got the same error condition - even though the said missing file was right there inside the directory:

------------------------------

---------------------
Begin Evaluation  198
---------------------
(Asynchronous job 198 added to queue)
First pass: initiating 1 local asynchronous jobs
Initiating evaluation 198
start "SystemInterface-Evaluation" /B simulator_script.bat params.in results.out

E:\NOTES\CYCLING NEW\dakota_cycling_Windows\main1\runOptim\workdir.198>echo off
E:\NOTES\CYCLING NEW\dakota_cycling_Windows\main1\runOptim\workdir.183>Evaluation 183 has completed

Nonblocking synchronize of 15 asynchronous evaluations
Second pass: testing for completions (15 running)

---------------------
Begin Evaluation  199
---------------------
(Asynchronous job 199 added to queue)
First pass: initiating 1 local asynchronous jobs
Initiating evaluation 199
start "SystemInterface-Evaluation" /B simulator_script.bat params.in results.out

E:\NOTES\CYCLING NEW\dakota_cycling_Windows\main1\runOptim\workdir.199>echo off

Error: cannot open results file "workdir.184\results.out" for evaluation 184

E:\NOTES\CYCLING NEW\dakota_cycling_Windows\main1\runOptim\workdir.184>
Musa@MLAHome MINGW64 /e/NOTES/CYCLING NEW/dakota_cycling_Windows/main1/runOptim
$
E:\NOTES\CYCLING NEW\dakota_cycling_Windows\main1\runOptim\workdir.186>
E:\NOTES\CYCLING NEW\dakota_cycling_Windows\main1\runOptim\workdir.185>
E:\NOTES\CYCLING NEW\dakota_cycling_Windows\main1\runOptim\workdir.187>
E:\NOTES\CYCLING NEW\dakota_cycling_Windows\main1\runOptim\workdir.188>
....
---------------------------------------
I checked in directory workdir.184 and the file results.out is right in there! Any ideas?
Thanks.
Musa
